### PR TITLE
Enable autovacuum, to deal with 'template0'

### PR DESCRIPTION
**Pull request by hlinnaka**
There are two commits here. The first one, to track "oldest xmin" in segments, is perhaps the  more interesting one. See commit messages, for details.

This currently fails the `vacuum_full_recently_dead_tuple_due_to_distributed_snapshot`, but that's OK: The test was written to test that VACUUM in utility mode is conservative, and won't remove tuples that are in fact dead. But with this PR, it can do so safely. So the test tested the overly-conservative approach that it used to do. We should probably change it or write a new one, to test that VACUUM in utility mode doesn't incorrectly remove tuples that are still visible to a distributed snapshot.

### CODE REVIEW

**Comment by hlinnaka** [comment]
Oh, and the third commit adds a GUC that I found really useful while testing this. To rack up the XID counter in a segment, I used `PGOPTIONS='-c gp_session_role=utility' PGPORT=40000 pgbench -f burn-xids.sql -n postgres`, with this `burn-xids.sql` script:
```
set debug_burn_xids=on;
select txid_current();
```

File: src/backend/access/transam/distributedlog.c [hunk 1]
```diff
@@ -73,6 +77,138 @@ static bool DistributedLog_PagePrecedes(int page1, int page2);
...
+	{
+		if (TransactionIdEquals(xid, oldestLocalXmin))
+			break;
+
+		int			page = TransactionIdToPage(xid);
+		int			entryno = TransactionIdToEntry(xid);
+
+		if (page != currPage)
+		{
+			slotno = SimpleLruReadPage(DistributedLogCtl, page, true, xid);
```
Comment (ashwinstar) [inline src/backend/access/transam/distributedlog.c]: We should leverage the process local cache here `LocalDistribXactCache_CommittedFind()` to help find localXid to distributedXid mapping faster if possible.

File: src/backend/access/transam/distributedlog.c [hunk 2]
```diff
@@ -73,6 +77,138 @@ static bool DistributedLog_PagePrecedes(int page1, int page2);
...
+
+	LWLockAcquire(DistributedLogControlLock, LW_EXCLUSIVE);
+
+	oldestXmin = DistributedLogShared->oldestXmin;
+
+	if (oldestXmin == InvalidTransactionId)
+	{
+		/*
+		 * If we haven't "seen" any distributed snapshots from the QD yet,
+		 * initialize oldest xmin to the oldest datfrozenxid in this segment.
```
Comment (ashwinstar) [inline src/backend/access/transam/distributedlog.c]: ya starting from `datfrozenxid` seems too much processing, must preserve the same across restarts.

File: src/backend/access/transam/distributedlog.c [hunk 3]
```diff
@@ -73,6 +77,138 @@ static bool DistributedLog_PagePrecedes(int page1, int page2);
...
+		 * call, the loop below can wind through millions of XIDs.
+		 * Fortunately, that only happens on the first call. But if that
+		 * ever becomes a problem, we could persist the value across
+		 * server restarts e.g. in the checkpoint record.
+		 *
+		 * We must release DistributedLogControlLock first, to avoid deadlock.
+		 * (GetNewTransactionId() can call DistributedLog_Truncate(), while
+		 * already holding XidGenLock)
+		 */
+		LWLockRelease(DistributedLogControlLock);
```
Comment (ashwinstar) [inline src/backend/access/transam/distributedlog.c]: Not fan of `LWLockRelease()` and `LWLockAcquire()` again, does it pose race here where other backend updates the `DistributedLogShared->oldestXmin` meanwhile. Seems mostly would just do duplicate work ?

File: src/backend/access/transam/distributedlog.c [hunk 4]
```diff
@@ -73,6 +77,138 @@ static bool DistributedLog_PagePrecedes(int page1, int page2);
...
+		}
+
+		ptr = &entries[entryno];
+
+		/*
+		 * If this XID is already visible to all distributed snapshots, we can advance
+		 * past it. Otherwise stop here.
+		 */
+		if (ptr->distribTimeStamp != distribTransactionTimeStamp ||
+			TransactionIdPrecedes(ptr->distribXid, xminAllDistributedSnapshots))
```
Comment (ashwinstar) [inline src/backend/access/transam/distributedlog.c]: The approach seems lot more conservative but maybe fine in practice, as higher localXid can map to lower distribXid which can reclaimed but it stops here.

File: src/backend/storage/ipc/procarray.c [hunk 5]
```diff
@@ -1857,6 +1891,20 @@ GetSnapshotData(Snapshot snapshot)
 	if (TransactionIdPrecedes(xmin, globalxmin))
 		globalxmin = xmin;
 
+	/*
+	 * In computing RecentGlobalXmin, also take distributed snapshots into
+	 * account.
+	 */
+	if (Gp_segment != -1)
+	{
+		DistributedSnapshot *ds = &snapshot->distribSnapshotWithLocalMapping.ds;
```
Comment (ashwinstar) [inline src/backend/storage/ipc/procarray.c]: need validation if DistributedSnapshot is not present ?

File: src/backend/access/transam/distributedlog.c [hunk 6]
```diff
@@ -73,6 +77,138 @@ static bool DistributedLog_PagePrecedes(int page1, int page2);
...
+								 DistributedTransactionId xminAllDistributedSnapshots)
+{
+	TransactionId xid;
+	TransactionId oldestXmin;
+	int			currPage;
+	int			slotno;
+	DistributedLogEntry *ptr;
+	DistributedLogEntry *entries;
+
+	LWLockAcquire(DistributedLogControlLock, LW_EXCLUSIVE);
```
Comment (ashwinstar) [inline src/backend/access/transam/distributedlog.c]: Seems will introduce contention on `DistributedLogControlLock`.

File: src/backend/access/transam/distributedlog.c (hunk 1 above)
Comment (hlinnaka) [reply src/backend/access/transam/distributedlog.c]: Hmm, I don't think that's useful here. I suppose we could use it, instead of reading the SLRU page, if the value happens to be in the cache. But: if there are any XIDs that are not found in the cache, we'd have to hit the SLRU anyway. And because we're scanning the SLRU sequentially, it's faster than using the cache, if we can advance the value more than 1-2 xids. And we'll need to grab the DistributedLogControlLock anyway, in order to update the oldestxmin value kept in shared memory.

File: src/backend/access/transam/distributedlog.c (hunk 3 above)
Comment (hlinnaka) [reply src/backend/access/transam/distributedlog.c]: Yeah, it's just duplicated work. I'll add a check, after re-acquiring the lock, to see if `DistributedLogShared->oldestXmin` was changed, while we were not holding the lock. That'll avoid scanning the SLRU, if another backend already did it.

File: src/backend/access/transam/distributedlog.c (hunk 4 above)
Comment (hlinnaka) [reply src/backend/access/transam/distributedlog.c]: Yeah. That is the general tradeoff we make with OldestXmin. It's also possible with purely local XIDs, that OldestXmin is held back by one very old transaction, and we could safely remove/freeze some XIDs newer than OldestXmin, if we just kept track of snapshots in a more fine-grained fashion. HeapTupleSatisfiesVacuum() is called for every tuple during VACUUM, so it's important that the test against OldestXmin is simple and cheap.

File: src/backend/access/transam/varsup.c [hunk 7]
```diff
@@ -186,6 +186,21 @@ GetNewTransactionId(bool isSubXact)
 	 */
 	TransactionIdAdvance(ShmemVariableCache->nextXid);
 
+	if (Debug_burn_xids)
```
Comment (xinzweb) [inline src/backend/access/transam/varsup.c]: Shall we do the `Debug_burn_xids` before the `TransactionIdAdvance` so that it can wrap around?

File: src/backend/access/transam/varsup.c [hunk 8]
```diff
@@ -186,6 +186,21 @@ GetNewTransactionId(bool isSubXact)
...
+	{
+		TransactionId xx;
+		int			r;
+
+		xx = ShmemVariableCache->nextXid;
+
+		r = xx % 1024;
```
Comment (xinzweb) [inline src/backend/access/transam/varsup.c]: Not sure the logic using 1024, why not using `CLOG_XACTS_PER_PAGE`? Then, if we turn the `Debug_burn_xids`, each transaction will jump to end of the current clog page?

File: src/backend/access/transam/varsup.c [hunk 9]
```diff
@@ -186,6 +186,21 @@ GetNewTransactionId(bool isSubXact)
 	 */
 	TransactionIdAdvance(ShmemVariableCache->nextXid);
 
+	if (Debug_burn_xids)
+	{
+		TransactionId xx;
+		int			r;
```
Comment (xinzweb) [inline src/backend/access/transam/varsup.c]: Why not just use unit32 for the r? Under what condition, do we want this r to be negative? Thanks.

**Comment by xinzweb** [comment]
@hlinnaka add test to verify autovacuum both positive and negative test cases like:
- autovacuum only kicked in for the `template0`
- autovacuum not triggered on `tempate1` and any other user databases like `postgres`

File: src/backend/postmaster/autovacuum.c [hunk 10]
```diff
@@ -1803,6 +1802,16 @@ get_database_list(void)
 		Form_pg_database pgdatabase = (Form_pg_database) GETSTRUCT(tup);
 		avw_dbase   *avdb;
 
+		/*
+		 * In GPDB, autovacuum is currently disabled, except for the
```
[the remark this answers is not in the document]
Comment (xinzweb) [reply src/backend/postmaster/autovacuum.c]: One possible solution would skip the shared relations when building the `table_oids` inside `do_autovacuum()`? In this case, we will not touch any shared relations and make this autovacuum fully localize to that give database (template0).

**Comment by hlinnaka** [comment]
Pushed a new version, with some changes to the way the "OldestXmin" value in shared memory is initialized. I think this should address Ashwin's performance concerns on the first call.

On the first call after postmaster startup, like before, we now start scanning from the oldest datfrozenxid. But instead of scanning one XID at a time, we first advance one SLRU page at a time with 'SimpleLruPageExists`, skipping over pg_distributedlog pages that have already been truncated away. And to make sure that helps, we truncate pg_distributedlog more aggressively, whenever we find that oldestXmin has advanced to a new SLRU segment. That should greatly speed up the initialization after postmaster startup.

This doesn't change the fact that we now need to acquire DistributedLogControlLock every time a snapshot is acquired. That could become a problem, hurting concurrency. To alleviate that, we could make it work so that only one backend would try to advance the oldest xmin value at a time, and if the lock is taken because someone else is already advancing it, we would just use the old value.  But I'd like to keep this simple, until someone demonstrates that there is an actual problem. These things can easily be masked by much bigger overheads elsewhere.

**Comment by ashwinstar** [comment]
On Tue, Mar 6, 2018 at 6:52 AM, Heikki Linnakangas <notifications@github.com
> wrote:

> Pushed a new version, with some changes to the way the "OldestXmin" value
> in shared memory is initialized. I think this should address Ashwin's
> performance concerns on the first call.
>
> On the first call after postmaster startup, like before, we now start
> scanning from the oldest datfrozenxid. But instead of scanning one XID at a
> time, we first advance one SLRU page at a time with 'SimpleLruPageExists`,
> skipping over pg_distributedlog pages that have already been truncated
> away. And to make sure that helps, we truncate pg_distributedlog more
> aggressively, whenever we find that oldestXmin has advanced to a new SLRU
> segment. That should greatly speed up the initialization after postmaster
> startup.
>
Lovely !!! Really liked the simplicity of avoiding the performance penalty
by aggressive distributed log truncation.
Actually, it may turn out to be performance boost for regular queries as
well, as might result in faster dxid->lxid lookups.

> This doesn't change the fact that we now need to acquire
> DistributedLogControlLock every time a snapshot is acquired. That could
> become a problem, hurting concurrency. To alleviate that, we could make it
> work so that only one backend would try to advance the oldest xmin value at
> a time, and if the lock is taken because someone else is already advancing
> it, we would just use the old value. But I'd like to keep this simple,
> until someone demonstrates that there is an actual problem. These things
> can easily be masked by much bigger overheads elsewhere.
>
Tricky part is since there exist no performance pipeline to measure and
validate the performance regressions, someone else will demonstrate the
problem seems unlikely :-) Will trying just some simple pg_bench
concurrency runs will give some rough idea, if this affects or not ?

**Comment by ashwinstar** [comment]
Just thinking is there possibility to land in weird conditions for incorrectly missing distributed log files, as now its not treated as ERROR condition but rather driving decisions ?

File: src/backend/access/transam/distributedlog.c [hunk 11]
```diff
@@ -69,10 +93,207 @@ static void DistributedLog_SetCommitted(TransactionId localXid,
...
+	DistributedLogShared->oldestXmin = oldestXmin;
+
+	/*
+	 * Done. But if we crossed an SLRU segment boundary, we can now remove
+	 * some old segments.
+	 */
+	if (oldOldestXmin % (ENTRIES_PER_PAGE * SLRU_PAGES_PER_SEGMENT) !=
+		oldestXmin % (ENTRIES_PER_PAGE * SLRU_PAGES_PER_SEGMENT))
+	{
+		DistributedLog_Truncate();
```
Comment (asimrp) [inline src/backend/access/transam/distributedlog.c]: GPDB is notorious in accumulating distributed logs in large volumes.  Truncating distributed log so aggressively is, therefore, quite cool!

It's ok to truncate distributed log without freezing the tuples because:
1. Non-existence of distributed log implies visibility is solely determined by local XID.
AND
2. The QE is guaranteed to receive no distributed snapshot whose in-prgress (or in-future) distributed XIDs fall into the range we just truncated.

If that makes sense, adding a comment as such will help.

**Comment by Eulerizeit** [comment]
I don't think we should hold this based on performance concerns. Aside from that, if we are sure we won't end in a weird state (Ashwin's question) then we should push.

File: src/backend/access/transam/distributedlog.c (hunk 11 above)
Comment (asimrp) [reply src/backend/access/transam/distributedlog.c]: `xminAllDistributedSnapshots` is monotonically increasing, it can never decrease.  That proves point 2 above.  Does that help ruling out Ashwin's concern?

File: src/backend/access/transam/distributedlog.c [hunk 12]
```diff
@@ -69,10 +93,207 @@ static void DistributedLog_SetCommitted(TransactionId localXid,
...
+	while (!TransactionIdEquals(xid, oldestLocalXmin))
+	{
+		int			page = TransactionIdToPage(xid);
+		int			entryno = TransactionIdToEntry(xid);
+
+		if (page != currPage)
+		{
+			slotno = SimpleLruReadPage(DistributedLogCtl, page, true, xid);
+			entries = (DistributedLogEntry *) DistributedLogCtl->shared->page_buffer[slotno];
+		}
```
[the remark this answers is not in the document]
Comment (hlinnaka) [reply src/backend/access/transam/distributedlog.c]: Oops, fixed.

**Comment by hlinnaka** [comment]
On 07/03/18 12:33, Asim RP wrote:
> The SLRU cache is meant to help backends querying commit status of
> distributed transactions in distributed log.  Using SLRU for
> sequential scans like this affects the LRU bookkeeping adversely, by
> possibly evicting "real recently used" pages unnecessarily.

Well, the LRU bookkeeping will keep the recently used pages in memory, 
regardless of what they're used for. Checking the page that corresponds 
to oldestxmin is just as valid a use of the SLRU as the other queries. I 
don't see a problem here.

(Without a concrete test case that shows a problem, or a concrete 
benchmark that shows that there is no problem, this is all just 
hand-waving.)

- Heikki

**Comment by hlinnaka** [comment]
On 06/03/18 20:59, Ashwin Agrawal wrote:
> Just thinking is there possibility to land in weird conditions for
> incorrectly missing distributed log files, as now its not treated as
> ERROR condition but rather driving decisions ?

Yeah, that's a possibility. Or rather, it's a "can't happen" scenario 
that will undoubtedly happen at some point :-).

In this PR, I have a PANIC, if it appears that the oldestXmin in shared 
memory goes past a locally computed oldestXmin. I'm thinking that before 
pushing this, we should downgrade that to an ERROR. In general, any 
confusion or errors around this code should be ephemeral, i.e it should 
fix itself once all the in-progress transactions and snapshots are gone. 
It's good if we can have safeguards to detect apparent confusion, and 
throw ERRORs, but the system as whole should recover.

- Heikki

**Comment by hlinnaka** [comment]
Squashed all the bug-fixes, and rebased. This is now ready to be committed, as far as I'm concerned, as soon as the PR pipeline is happy. Any last minute concerns?
